File: Difficulty: Medium/Next Greater Element in Circular Array/next-greater-element-in-circular-array.py
class Solution:
    def nextLargerElement(self, arr):
        n = len(arr)
        res = [-1] * n
        stack = []

        # Traverse the array twice to simulate circular nature
        for i in range(2 * n - 1, -1, -1):
            index = i % n
            while stack and arr[stack[-1]] <= arr[index]:
                stack.pop()
            if i < n and stack:
                res[index] = arr[stack[-1]]
            stack.append(index)

        return res

File: Difficulty: Medium/Next Greater Element in Circular Array/next-greater-element-in-circular-array.java
import java.util.*;

class Solution {
    public ArrayList<Integer> nextGreater(int[] arr) {
        int n = arr.length;
        ArrayList<Integer> res = new ArrayList<>(Collections.nCopies(n, -1));
        Stack<Integer> st = new Stack<>();
        
        for (int i = 0; i < 2 * n; i++) {
            int curr = arr[i % n];
            while (!st.isEmpty() && arr[st.peek()] < curr) {
                res.set(st.pop(), curr);
            }
            if (i < n) {
                st.push(i);
            }
        }
        return res;
    }
}
